### ISSUE TITLE
Missing parameters: client_id during obtaining accessToken using refreshToken

### ISSUE THREAD

**Issue by jack-fdrv** [upwork/node-upwork-oauth2#29]
Looks like we are no more able to use refreshToken and get accessToken via setAccessToken it returns this one

```
Uncaught Exception: FDRV AAAAA Error: Response Error: 400 Bad Request
    at internals.Client._shortcut (fdrv/node_modules/@hapi/wreck/lib/index.js:569:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Client.request (fdrv/node_modules/simple-oauth2/lib/client/client.js:39:22)
    at async AccessToken.refresh (fdrv/node_modules/simple-oauth2/lib/access-token.js:49:22) {
  data: {
    isResponseError: true,
    headers: {
      date: 'Sat, 23 Mar 2024 21:58:18 GMT',
      'content-type': 'application/json',
      'content-length': '79',
      connection: 'close',
      'vnd.eo.request-id': 'd1e7b03e-1283-4ff8-9ae6-f4a28a3aaa03',
      'vnd.odesk.request-id': 'd1e7b03e-1283-4ff8-9ae6-f4a28a3aaa03',
      'vnd-eo-trace-id': '8691ae9b9f099f62-AMS, 8691ae9b9f099f62-AMS',
      'vnd-eo-span-id': '3331a322-fdf5-4b8c-bac9-2d47d1a77fc4',
      'vnd-eo-parent-span-id': '8691ae9b9f099f62-AMS',
      'vnd.eo.trace-id': '8691ae9b9f099f62-AMS',
      'vnd.eo.span-id': '3331a322-fdf5-4b8c-bac9-2d47d1a77fc4',
      'vnd.eo.parent-span-id': '8691ae9b9f099f62-AMS',
      'x-upwork-target-status': 'true',
      'vnd-eo-server-sr': '2024-03-23T21:58:18.301Z',
      'vnd-eo-server-ss': '2024-03-23T21:58:18.305Z',
      'vnd-eo-prana-client-receive': '2024-03-23T21:58:18.305Z, 2024-03-23T21:58:18.306Z',
      'referrer-policy': 'origin-when-cross-origin',
      'x-xss-protection': '1; mode=block',
      'cf-cache-status': 'DYNAMIC',
      'set-cookie': [Array],
      'strict-transport-security': 'max-age=31536000; includeSubDomains; preload',
      'x-content-type-options': 'nosniff',
      server: 'cloudflare',
      'cf-ray': '8691ae9b9f099f62-AMS',
      'alt-svc': 'h3=":443"; ma=86400'
    },
    res: IncomingMessage {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      socket: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      rawHeaders: [Array],
      rawTrailers: [],
      joinDuplicateHeaders: undefined,
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 400,
      statusMessage: 'Bad Request',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [ClientRequest],
      [Symbol(kCapture)]: false,
      [Symbol(kHeaders)]: [Object],
      [Symbol(kHeadersCount)]: 58,
      [Symbol(kTrailers)]: null,
      [Symbol(kTrailersCount)]: 0
    },
    payload: {
      error_description: 'Missing parameters: client_id',
      error: 'invalid_request'
    }
  },
  isBoom: true,
  isServer: false,
  output: {
    statusCode: 400,
    payload: {
      statusCode: 400,
      error: 'Bad Request',
      message: 'Response Error: 400 Bad Request'
    },
    headers: {}
  }
}

```

actually from docs we dont have client_id we do have clientId

```
var config = {
  'clientId' : 'xxxxxxxx',
  'clientSecret' : 'xxxxxxxx',
  'redirectUri': 'xxxxxxxx';
```
  
  so then we do 
  
`  var api = new UpworkApi(config);`

and then if we have refreshToken we do 

`api.setAccessToken(function(tokenPair) {`

and it fails.

**Comment by mnovozhylov** [comment]
hi @jack-fdrv 
`clientId` is transformed to `client_id` in the oauth2 library. And the response above comes from the server side.
Have you tried a raw example before integrating into your application - does it work for you? I'd also debug the `tokenPair` value.

**Comment by dmatora** [comment]
@mnovozhylov 
same issue here, error_description: 'Missing parameters: client_id'
that's output of example/example.js

php library works fine

**Comment by dmatora** [comment]
Here is workaround to get refresh token working using TypeScript

```
const refreshAccessToken = async (
  refreshToken: string
): Promise<null | {
  access_token: string;
  refresh_token: string;
  expires_in: number;
}> => {
  try {
    const data = new URLSearchParams({
      client_id: UPWORK_API_KEY,
      client_secret: UPWORK_API_SECRET,
      refresh_token: refreshToken,
      grant_type: 'refresh_token',
    }).toString();

    const config = {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    };

    const response = await axios.post(
      'https://www.upwork.com/api/v3/oauth2/token',
      data,
      config
    );

    if (response.status === 200) {
      return { ...response.data, token_type: undefined };
    } else {
      console.log(
        'Failed to refresh access token. Status code:',
        response.status,
        response.statusText
      );
    }
  } catch (error) {
    console.error('Error refreshing access token:', error);
  }
  return null;
};
```

**Comment by jack-fdrv** [comment]
> Here is workaround to get refresh token working using TypeScript
> 
> ```
> const refreshAccessToken = async (
>   refreshToken: string
> ): Promise<null | {
>   access_token: string;
>   refresh_token: string;
>   expires_in: number;
> }> => {
>   try {
>     const data = new URLSearchParams({
>       client_id: UPWORK_API_KEY,
>       client_secret: UPWORK_API_SECRET,
>       refresh_token: refreshToken,
>       grant_type: 'refresh_token',
>     }).toString();
> 
>     const config = {
>       headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
>     };
> 
>     const response = await axios.post(
>       'https://www.upwork.com/api/v3/oauth2/token',
>       data,
>       config
>     );
> 
>     if (response.status === 200) {
>       return { ...response.data, token_type: undefined };
>     } else {
>       console.log(
>         'Failed to refresh access token. Status code:',
>         response.status,
>         response.statusText
>       );
>     }
>   } catch (error) {
>     console.error('Error refreshing access token:', error);
>   }
>   return null;
> };
> ```

Works. Thanks. God Bless you Dmitry =)

**Comment by dmatora** [comment]
Workaround doesn't refresh token automatically
Created PR with a proper fix

**Comment by mnovozhylov** [comment]
Hi @dmatora 
Thank you for your contribution! The fix has been accepted and delivered in `v2.2.2`.

**Comment by hadnazzar** [comment]
@mnovozhylov @dmatora when token expired that fix does not refreshes the token again. 

How to reproduce? 
- Generate token
- After a 24 hours token expires
- try to run with same token 
- Getting error like below: 
```
{
  "error_description": "Refresh token is expired or revoked or not registered",
  "error": "invalid_grant"
}
```
